File: tests/oracle/deindirect_arms_1204.cc
/*
 * DEINDIRECT-ARMS-1204: locked Ghidra 12.0.4 oracle for
 * FSPEC-DEINDIRECT-TRIGGER-0001 — the three conversion arms of
 * ActionDeindirect::apply (coreaction.cc:1219-1280) and the observable
 * slice of the FuncCallSpecs conversions they drive
 * (fspec.cc:5443-5472 deindirect / fspec.cc:5485-5509 forceSet).
 *
 * Covered projection (all observations through public Ghidra API plus the
 * two fixture-only accessors the runner patches into the throwaway
 * archive): the CALLIND input(0) COPY-chain walk (cc:1231-1232), the
 * external-reference arm (cc:1233-1240, queryExternalRefFunction via the
 * real global scope), the constant arm with the funcptr_align encoding-bit
 * strip (cc:1241-1257, AddrSpace::addressToByte + arch->funcptr_align),
 * the isOverride early return inside deindirect (fspec.cc:5462), the
 * noreturn/inline restart gate (fspec.cc:5461/5471), and the
 * typed-function-pointer forceSet arm (cc:1258-1277,
 * getTypeReadFacing PTR->CODE + TypeCode prototype + isInputLocked).
 *
 * Host: the cptr_b_1204 BfdArchitecture pattern (examples/curl image, the
 * GetStr function), with query targets installed on the real global scope
 * through public Scope::addFunction / Scope::addExternalRef at image-safe
 * high addresses.
 */

#include "bfd_arch.hh"
#include "coreaction.hh"
#include "fspec.hh"
#include "funcdata.hh"
#include "libdecomp.hh"
#include "type.hh"

#include <iomanip>
#include <iostream>
#include <map>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

using namespace ghidra;
using std::map;
using std::ostringstream;
using std::string;
using std::vector;

class InspectableDeindirect : public ActionDeindirect {
public:
  InspectableDeindirect(void) : ActionDeindirect("deindirect") {
    // Action::Action leaves count/lcount uninitialized; perform()
    // normally initializes them. This fixture calls apply() directly.
    count = 0;
    lcount = 0;
  }
  int4 fixtureCount(void) const { return count; }
};

class Site {
public:
  Funcdata &fd;
  PcodeOp *op;
  FuncCallSpecs *fc;

  Site(Funcdata &d, AddrSpace *ram, uintb callpc)
    : fd(d), op((PcodeOp *)0), fc((FuncCallSpecs *)0)
  {
    op = fd.newOp(1, Address(ram, callpc));
    fd.opSetOpcode(op, CPUI_CALLIND);
    fc = new FuncCallSpecs(op);
    // Production reality (coreaction.cc:2311-2332): every callspec has
    // passed ActionDefaultParams by the time deindirect runs, so it always
    // carries a model and a proto store; the fixture installs the same
    // setInternal(defaultfp, void) shape (a bare FuncProto has a null
    // store, which the dump's isInputLocked read would dereference).
    fc->setInternal(d.getArch()->defaultfp, d.getArch()->types->getTypeVoid());
    fd.opSetInput(op, fd.newVarnodeCallSpecs(fc), 0);
    fd.fixtureAddToCallList(fc);
    BlockBasic *block = const_cast<BlockGraph &>(fd.getBasicBlocks()).newBlockBasic(&fd);
    fd.opInsertEnd(op, block);
  }

  /// The deindirect observable slice after apply: opcode number (the enum
  /// value is the shared contract with the Rust side), entry address
  /// validity/offset, the adopted display name, the restart flag DELTA
  /// (setRestartPending is Funcdata-global and sticky, so the per-case
  /// observable is "this apply fired a restart"), the input-lock state,
  /// the override flag, and the CALL's input arity.
  void dump(const string &caseName, int4 count, Funcdata &data, bool restartBefore)
  {
    ostringstream out;
    out << "case=" << caseName
        << "|op=" << static_cast<int4>(op->code())
        << "|entry=";
    if (fc->getEntryAddress().isInvalid())
      out << "inv";
    else
      out << "0x" << std::hex << fc->getEntryAddress().getOffset() << std::dec;
    bool restartFired = data.hasRestartPending() && !restartBefore;
    out << "|name=" << fc->getName()
        << "|count=" << count
        << "|restart=" << (restartFired ? 1 : 0)
        << "|inlock=" << (fc->isInputLocked() ? 1 : 0)
        << "|override=" << (fc->isOverride() ? 1 : 0)
        << "|arity=" << op->numInput();
    std::cout << out.str() << '\n';
    std::cout.flush();
  }
};

void runCases(Architecture *glb, Funcdata *fdp)
{
  Funcdata &fd = *fdp;
  AddrSpace *ram = glb->getDefaultCodeSpace();
  Scope *globals = glb->symboltab->getGlobalScope();

  // ---- case const_hit: constant arm resolves through queryFunction ----
  globals->addFunction(Address(ram, 0x60000000), "target_const");
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500010);
    Varnode *cnst = fd.newConstant(8, 0x60000000);
    fd.opSetInput(site.op, cnst, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("const_hit", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case const_miss: no function at the constant address ----
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500020);
    Varnode *cnst = fd.newConstant(8, 0x60001000);
    fd.opSetInput(site.op, cnst, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("const_miss", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case align_strip: funcptr_align strips the encoding bits ----
  // cc:1245-1249: funcptr_align=2 -> offset >>= 2; offset <<= 2 before the
  // query, so the low bits never reach queryFunction.
  glb->funcptr_align = 2;
  globals->addFunction(Address(ram, 0x60002000), "target_aligned");
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500030);
    Varnode *cnst = fd.newConstant(8, 0x60002003);
    fd.opSetInput(site.op, cnst, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("align_strip", action.fixtureCount(), fd, restartBefore);
  }
  glb->funcptr_align = 0;

  // ---- case copy_chain: input(0) resolves through a COPY chain ----
  // cc:1231-1232: the walk follows COPY defs; the arms examine the walked
  // varnode (the constant), not the original COPY output.
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500040);
    Varnode *cnst = fd.newConstant(8, 0x60000000);
    PcodeOp *copyop = fd.newOp(1, Address(ram, 0x500041));
    fd.opSetOpcode(copyop, CPUI_COPY);
    Varnode *copyout = fd.newUniqueOut(8, copyop);
    fd.opSetInput(copyop, cnst, 0);
    fd.opInsertBefore(copyop, site.op);
    fd.opSetInput(site.op, copyout, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("copy_chain", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case noreturn: the fspec.cc:5461 gate falls to setRestartPending ----
  globals->addFunction(Address(ram, 0x60003000), "t_noreturn");
  {
    Funcdata *callee = globals->queryFunction(Address(ram, 0x60003000));
    callee->getFuncProto().setNoReturn(true);
  }
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500050);
    Varnode *cnst = fd.newConstant(8, 0x60003000);
    fd.opSetInput(site.op, cnst, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("norestart_gate", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case override_site: fspec.cc:5462 early return (no restart, no
  //      late restriction) on a callspec carrying an applied override ----
  globals->addFunction(Address(ram, 0x60004000), "t_override");
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500060);
    Varnode *cnst = fd.newConstant(8, 0x60004000);
    fd.opSetInput(site.op, cnst, 0);
    site.fc->setOverride(true);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("override_site", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case extref: external-reference arm (cc:1233-1240) ----
  // The CALLIND input is a persist+externref varnode whose address holds an
  // ExternRefSymbol referring to the real function.
  globals->addFunction(Address(ram, 0x60006000), "ext_target");
  globals->addExternalRef(Address(ram, 0x60005000), Address(ram, 0x60006000), "ex_slot");
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500070);
    Varnode *exvn = fd.newVarnode(8, Address(ram, 0x60005000));
    exvn->fixtureSetExternRef();
    fd.opSetInput(site.op, exvn, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("extref", action.fixtureCount(), fd, restartBefore);
  }

  // ---- case funcptr_force: typed-funcptr forceSet arm (cc:1258-1277) ----
  // Type recovery started + input(0) typed PTR->CODE with an attached
  // prototype + callspec not input-locked -> forceSet: prototype override
  // recorded, prototype locked, arity collapses per commitNewInputs.
  fd.startTypeRecovery();
  {
    InspectableDeindirect action;
    Site site(fd, ram, 0x500080);
    Varnode *fpvn = fd.newVarnode(8, Address(ram, 0x60007000));
    PrototypePieces sig;
    sig.model = glb->defaultfp;
    sig.name = "fp_callee";
    sig.outtype = glb->types->getTypeVoid();
    sig.firstVarArgSlot = -1;
    TypeCode *tc = glb->types->getTypeCode(sig);
    TypePointer *ptr = glb->types->getTypePointer(8, tc, 1);
    fpvn->updateType(ptr, true, true);
    fd.opSetInput(site.op, fpvn, 0);
    bool restartBefore = fd.hasRestartPending();
    action.apply(fd);
    site.dump("funcptr_force", action.fixtureCount(), fd, restartBefore);
  }
}

void run(const string &specDirectory, const string &binary)
{
  vector<string> specPaths;
  specPaths.push_back(specDirectory);
  startDecompilerLibrary(specPaths);
  {
    BfdArchitecture architecture(binary, "default", &std::cerr);
    DocumentStorage store;
    architecture.init(store);
    architecture.readLoaderSymbols("::");
    Funcdata *fd = architecture.symboltab->getGlobalScope()->queryFunction("GetStr");
    if (fd == (Funcdata *)0)
      throw std::runtime_error("GetStr was not found in the BFD symbol table");
    if (fd->getName() != "GetStr" ||
        fd->getAddress().getOffset() != 0x36d0 || fd->getSize() != 0)
      throw std::runtime_error("GetStr input identity drifted");
    if (architecture.archid != "x86:LE:64:default:gcc")
      throw std::runtime_error("runtime architecture/compiler drifted: " +
                               architecture.archid);
    runCases(&architecture, fd);
  }
  shutdownDecompilerLibrary();
}

} // anonymous namespace

int main(int argc, char **argv)
{
  if (argc != 3) {
    std::cerr << "usage: deindirect_arms_1204 SPEC_ROOT CURL_BINARY\n";
    return 2;
  }
  try {
    run(argv[1], argv[2]);
    return 0;
  }
  catch (const LowlevelError &error) {
    std::cerr << "deindirect_arms_1204: LowlevelError: " << error.explain << '\n';
    return 1;
  }
  catch (const std::exception &error) {
    std::cerr << "deindirect_arms_1204: " << error.what() << '\n';
    return 1;
  }
}

File: tests/oracle/deindirect_arms_1204.rs
//! DEINDIRECT-ARMS-1204: Rust comparand for the locked Ghidra 12.0.4
//! oracle `deindirect_arms_1204.cc` (FSPEC-DEINDIRECT-TRIGGER-0001).
//!
//! Mirrors each case through the crate's public API: the real
//! `ActionDeindirect::apply` (coreaction.rs, coreaction.cc:1219-1280) over
//! a Funcdata whose Architecture carries the query-channel Database
//! (FunctionSymbols installed with `Scope::add_function`), so the constant
//! arm resolves through the same global-scope queryFunction production
//! uses. The two known production-channel residuals print in the same
//! observation format so the bilateral diff pinpoints them:
//!   - extref: Scope-side refaddr storage (CALLSPEC-0001 seam) — the
//!     detection runs but the referral cannot resolve, exactly like
//!     oracle's `newfd == 0` (no conversion);
//!   - norestart_gate: per-callee noreturn lives in the flow-time
//!     callee-proto channel only; the db-function observable slice Rugra
//!     mints at deindirect time cannot carry it yet.

use std::sync::{Arc, RwLock};

use rugra::address::Address;
use rugra::arch::Architecture;
use rugra::action::Action;
use rugra::coreaction::ActionDeindirect;
use rugra::database::Database;
use rugra::funcdata::Funcdata;
use rugra::fspec::{FuncCallSpecs, FuncProto};
use rugra::op::PcodeOpRef;
use rugra::opcodes::OpCode;
use rugra::type_system::datatype::{Datatype, TypeBase, TypeCode, TypeMetatype, TypePointer};
use rugra::varnode::varnode_flags;

fn void_type() -> Arc<Datatype> {
    Arc::new(Datatype::Void(TypeBase::new(
        "void".to_string(),
        0,
        TypeMetatype::Void,
    )))
}

/// One CALLIND call site: fresh op + callspec + typed annotation input,
/// the flow.cc:684-686 construction pattern.
struct Site {
    op: PcodeOpRef,
    owner: Arc<RwLock<FuncCallSpecs>>,
}

fn make_site(fd: &mut Funcdata, call_pc: u64) -> Site {
    let op = fd.new_op(1, Address::new(call_pc));
    fd.op_set_opcode(&op, OpCode::CPUI_CALLIND);
    // FuncCallSpecs::new_for_op's second parameter is the default proto
    // base; the fixture constructs the same default-constructed shape the
    // flow path uses (fspec.cc:4924 FuncProto() base).
    let proto = FuncProto::new(String::new(), void_type());
    let fc = FuncCallSpecs::new_for_op(&op, proto);
    let owner = Arc::new(RwLock::new(fc));
    fd.add_call_specs_owner(owner.clone());
    let annotation = fd.new_varnode_call_specs(&owner);
    fd.op_set_input(&op, annotation, 0);
    Site { op, owner }
}

/// The deindirect observable slice after apply, byte-mirroring the C++
/// Site::dump: opcode number (enum value contract), entry validity/offset,
/// adopted name, action count, restart flag, input lock, override flag,
/// CALL arity.
fn dump_site(fd: &Funcdata, case_name: &str, site: &Site, count: i32, restart_before: bool) {
    let (opcode, arity) = {
        let guard = site.op.0.read().unwrap();
        (guard.opcode as i32, guard.inrefs.len())
    };
    let fc = site.owner.read().unwrap();
    let entry = fc
        .entry_addr
        .map(|a| format!("0x{:x}", a.as_u64()))
        .unwrap_or_else(|| "inv".to_string());
    println!(
        "case={}|op={}|entry={}|name={}|count={}|restart={}|inlock={}|override={}|arity={}",
        case_name,
        opcode,
        entry,
        fc.prototype.name,
        count,
        (fd.restart_pending && !restart_before) as i32,
        fc.prototype.is_input_locked() as i32,
        fc.prototype.is_override() as i32,
        arity,
    );
}

fn site_arch(db_arc: &Arc<RwLock<Database>>, funcptr_align: i32) -> Arc<Architecture> {
    let mut arch = Architecture::new();
    arch.symboltab = Some(db_arc.clone());
    arch.funcptr_align = funcptr_align;
    Arc::new(arch)
}

fn main() {
    let db_arc = Arc::new(RwLock::new(Database::new(false)));
    {
        let mut db = db_arc.write().unwrap();
        let s = db.get_global_scope_mut().expect("global scope");
        s.add_function(Address::new(0x60000000), "target_const", 1);
        s.add_function(Address::new(0x60002000), "target_aligned", 1);
        s.add_function(Address::new(0x60003000), "t_noreturn", 1);
        s.add_function(Address::new(0x60004000), "t_override", 1);
        s.add_function(Address::new(0x60006000), "ext_target", 1);
    }

    let mut fd = Funcdata::new("GetStr", Address::new(0x36d0), 0);
    fd.set_arch(site_arch(&db_arc, 0));

    // ---- case const_hit: constant arm resolves through queryFunction ----
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500010);
        let cnst = fd.new_constant(8, 0x60000000);
        fd.op_set_input(&site.op, cnst, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply const_hit");
        dump_site(&fd, "const_hit", &site, action.count, restart_before);
    }

    // ---- case const_miss: no function at the constant address ----
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500020);
        let cnst = fd.new_constant(8, 0x60001000);
        fd.op_set_input(&site.op, cnst, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply const_miss");
        dump_site(&fd, "const_miss", &site, action.count, restart_before);
    }

    // ---- case align_strip: funcptr_align strips the encoding bits ----
    {
        let mut action = ActionDeindirect::new();
        fd.set_arch(site_arch(&db_arc, 2));
        let site = make_site(&mut fd, 0x500030);
        let cnst = fd.new_constant(8, 0x60002003);
        fd.op_set_input(&site.op, cnst, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply align_strip");
        dump_site(&fd, "align_strip", &site, action.count, restart_before);
        fd.set_arch(site_arch(&db_arc, 0));
    }

    // ---- case copy_chain: input(0) resolves through a COPY chain ----
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500040);
        let cnst = fd.new_constant(8, 0x60000000);
        let copyop = fd.new_op(1, Address::new(0x500041));
        fd.op_set_opcode(&copyop, OpCode::CPUI_COPY);
        let copyout = fd.new_unique_out(8, &copyop);
        fd.op_set_input(&copyop, cnst, 0);
        fd.op_insert_before(&copyop, &site.op);
        fd.op_set_input(&site.op, copyout, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply copy_chain");
        dump_site(&fd, "copy_chain", &site, action.count, restart_before);
    }

    // ---- case norestart_gate: fspec.cc:5461 noreturn callee -> restart.
    //      Production residual: the per-callee noreturn bit rides the
    //      flow-time callee-proto channel; the db observable slice Rugra
    //      mints here cannot carry it (registered on the ticket).
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500050);
        let cnst = fd.new_constant(8, 0x60003000);
        fd.op_set_input(&site.op, cnst, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply norestart_gate");
        dump_site(&fd, "norestart_gate", &site, action.count, restart_before);
    }

    // ---- case override_site: fspec.cc:5462 early return ----
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500060);
        let cnst = fd.new_constant(8, 0x60004000);
        fd.op_set_input(&site.op, cnst, 0);
        site.owner.write().unwrap().prototype.set_override(true);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply override_site");
        dump_site(&fd, "override_site", &site, action.count, restart_before);
    }

    // ---- case extref: external-reference arm. Production residual:
    //      Scope-side refaddr storage (CALLSPEC-0001 seam) — detection
    //      runs, referral cannot resolve (= oracle newfd==0, no
    //      conversion; the C++ fixture resolves through its real scope).
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500070);
        let exvn = fd
            .vbank
            .create_with_space(8, rugra::space::AddressSpace::Ram, 0x60005000);
        exvn.write().unwrap().flags |= varnode_flags::PERSIST | varnode_flags::EXTERNREF;
        fd.op_set_input(&site.op, exvn, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply extref");
        dump_site(&fd, "extref", &site, action.count, restart_before);
    }

    // ---- case funcptr_force: typed-funcptr forceSet arm (cc:1258-1277) ----
    fd.set_type_recovery_started();
    {
        let mut action = ActionDeindirect::new();
        let site = make_site(&mut fd, 0x500080);
        let fpvn = fd
            .vbank
            .create_with_space(8, rugra::space::AddressSpace::Ram, 0x60007000);
        let mut fp = FuncProto::new(String::new(), void_type());
        fp.set_internal(None, void_type());
        let mut tc = TypeCode::new();
        tc.proto = Some(Arc::new(fp));
        let code = Arc::new(Datatype::Code(tc));
        let ptr = Arc::new(Datatype::Pointer(TypePointer::new(8, code, 1)));
        fpvn.write().unwrap().v_type = Some(ptr);
        fd.op_set_input(&site.op, fpvn, 0);
        let restart_before = fd.restart_pending;
        action.apply(&mut fd).expect("apply funcptr_force");
        dump_site(&fd, "funcptr_force", &site, action.count, restart_before);
    }
}
